crates/ignore/src/walk.rs

```diff
@@ -1267,27 +1269,44 @@ impl WalkParallel {
                         }
                     }
                 };
-                stack.push(Message::Work(Work {
+                let is_dir = dent.is_dir();
+                let message = Message::Work(Work {
                     dent: dent,
                     ignore: self.ig_root.clone(),
                     root_device: root_device,
-                }));
+                });
+                if is_dir {
+                    directory_stack.push(message);
+                } else {
+                    file_stack.push(message);
+                }
             }
             // ... but there's no need to start workers if we don't need them.
-            if stack.is_empty() {
+            if file_stack.is_empty() && directory_stack.is_empty() {
                 return;
             }
         }
         // Create the workers and then wait for them to finish.
         let quit_now = Arc::new(AtomicBool::new(false));
-        let num_pending =
-            Arc::new(AtomicUsize::new(stack.lock().unwrap().len()));
+        let num_pending = Arc::new(AtomicUsize::new(
+            file_stack.lock().unwrap().len()
+                + directory_stack.lock().unwrap().len(),
+        ));
         crossbeam_utils::thread::scope(|s| {
             let mut handles = vec![];
-            for _ in 0..threads {
+            for i in 0..threads {
                 let worker = Worker {
                     visitor: builder.build(),
-                    stack: stack.clone(),
+                    // To avoid deadlock due to file stack size limit
+                    // some workers should be launched in consumer-only mode.
+                    // Testing shows that single producer has the best
+                    // performance but it can be tweaked if needed.
+                    directory_stack: if i == 0 {
```

Comment:
@BurntSushi this is the line that controls if the worker will have access to directories stack